### PR TITLE
Create a TRAPI interface

### LINKED ISSUE #27: Create a TRAPI interface
The interface would consume a TRAPI message, and would return a normalized TRAPI message.

### LINKED ISSUE #29: return types in the right format
Currently, types are being returned in snake case, like `chemical_substance`, but this needs to be fixed to return e.g. `biolink:ChemicalSubstance`

### LINKED ISSUE #31: Implement a POST
Currently we have only a GET interface.  We should also have a post.  The main reason is that GETs start to run into length issues.  So some part of the http stack at renci is complaining when the request header gets up to X bytes, which translates into something like a few hundred curies, after which it starts throwing 403.

With a POST, we can send much larger batches and provide some clearer guidance to callers.

### PR DESCRIPTION
**Pull request by kshefchek**
Goals:
- [x] refactor code to use Fast API
- [x] Add posts to get_normalized_nodes, get_curie_prefixes
- [x] Add function that takes a knowledge graph as input, normalizes nodes, and returns an updated and merged graph
- [x] Add endpoint that takes a TRAPI message as input and returns a normalized message
- [x] Update to TRAPI 1.0

closes #27 
closes #31 
could close #29, but wonder if this should be fixed in babel

### CODE REVIEW

**Comment by kshefchek** [comment]
This PR is ready for an initial review, so far I've implemented a basic knowledge merge.  Below is an example that I'm using as a test case in test_endpoints.py:

Premerged:
```json
{
  "nodes": [
    {
      "id": "HGNC:11603",
      "name": "TBX4",
      "type": [
        "biolink:Gene"
      ]
    },
    {
      "id": "NCBIGene:9496",
      "name": "T-box transcription factor 4",
      "type": [
        "biolink:Gene"
      ]
    },
    {
      "id": "MONDO:0005002",
      "name": "chronic obstructive pulmonary disease",
      "type": [
        "biolink:Disease"
      ]
    },
    {
      "id": "DOID:3083",
      "name": "chronic obstructive pulmonary disease",
      "type": [
        "biolink:Disease"
      ]
    },
    {
      "id": "OMIM:613355",
      "name": "CHROMOSOME 17q23.1-q23.2 DELETION SYNDROME",
      "type": [
        "biolink:Disease"
      ]
    },
    {
      "id": "UMLS:CN202575",
      "name": "heritable pulmonary arterial hypertension",
      "type": [
        "biolink:Disease"
      ]
    }
  ],
  "edges": [
    {
      "id": "a8575c4e-61a6-428a-bf09-fcb3e8d1644d",
      "source_id": "HGNC:11603",
      "target_id": "MONDO:0005002",
      "type": "biolink:related_to",
      "relation": "RO:0003304"
    },
    {
      "id": "b4971c4e-61a6-428a-bf09-fcb31234gfbs",
      "source_id": "NCBIGene:9496",
      "target_id": "DOID:3083",
      "type": "biolink:related_to",
      "relation": "RO:0003304"
    },
    {
      "id": "2d38345a-e9bf-4943-accb-dccba351dd04",
      "source_id": "NCBIGene:9496",
      "target_id": "OMIM:613355",
      "type": "biolink:related_to",
      "relation": "RO:0003304"
    },
    {
      "id": "044a7916-fba9-4b4f-ae48-f0815b0b222d",
      "source_id": "HGNC:11603",
      "target_id": "UMLS:CN202575",
      "type": "biolink:related_to",
      "relation": "RO:0004013"
    }
  ]
}
```

post-merged:
```json
{
  "nodes": [
    {
      "id": "NCBIGene:9496",
      "type": ["biolink:Gene"],
      "name": "TBX4",
      "same_as": ["NCBIGene:9496", "ENSEMBL:ENSG00000121075", "HGNC:11603", "UniProtKB:P57082"],
      "category": ["biolink:Gene", "biolink:NamedThing", "biolink:BiologicalEntity", "biolink:MolecularEntity", "biolink:GenomicEntity", "biolink:MacromolecularMachine", "biolink:GeneOrGeneProduct"]
    }, 
    {
      "id": "MONDO:0005002",
      "type": ["biolink:Disease"],
      "name": "chronic obstructive pulmonary disease",
      "same_as": ["MONDO:0005002", "DOID:3083", "EFO:0000341", "UMLS:C0024117", "MESH:D029424", "NCIT:C3199", "SNOMEDCT:13645005", "HP:0006510"], 
      "category": ["biolink:Disease", "biolink:NamedThing", "biolink:BiologicalEntity", "biolink:DiseaseOrPhenotypicFeature"]
    }, 
    {
      "id": "MONDO:0013238",
      "type": ["biolink:Disease"],
      "name": "chromosome 17q23.1-q23.2 deletion syndrome",
      "same_as": ["MONDO:0013238", "DOID:0060405", "OMIM:613355", "ORPHANET:261279", "UMLS:C3150607", "UMLS:C4304591", "SNOMEDCT:719584008"],
      "category": ["biolink:Disease", "biolink:NamedThing", "biolink:BiologicalEntity", "biolink:DiseaseOrPhenotypicFeature"]
    }, 
    {
      "id": "MONDO:0017148",
      "type": ["biolink:Disease"],
      "name": "heritable pulmonary arterial hypertension",
      "same_as": ["MONDO:0017148", "ORPHANET:275777", "UMLS:CN202575", "NCIT:C121945", "SNOMEDCT:697897003"],
      "category": ["biolink:Disease", "biolink:DiseaseOrPhenotypicFeature", "biolink:BiologicalEntity", "biolink:NamedThing"]
    }
  ],
  "edges": [
    {
      "id": "a8575c4e-61a6-428a-bf09-fcb3e8d1644d",
      "source_id": "NCBIGene:9496",
      "target_id": "MONDO:0005002",
      "type": "biolink:related_to",
      "relation": "RO:0003304"
    }, 
    {
      "id": "2d38345a-e9bf-4943-accb-dccba351dd04",
      "source_id": "NCBIGene:9496",
      "target_id": "MONDO:0013238",
      "type": "biolink:related_to",
      "relation": "RO:0003304"
    }, 
    {
      "id": "044a7916-fba9-4b4f-ae48-f0815b0b222d",
      "source_id": "NCBIGene:9496",
      "target_id": "MONDO:0017148",
      "type": "biolink:related_to",
      "relation": "RO:0004013"
    }
  ]
}
```

**Comment by cbizon** [comment]
This looks very nice, and the test case is good.  I have a few small requests for change -

I think we already decided this one on Friday, but just to document it we can leave out the camel case stuff - it'll be better fixed upstream (babel).

A little bigger, but hopefully not too bad, I think what we would like here is a full TRAPI interface, not just the knowledge graph component.  That's because if nodes change their identifiers in the KG, the node/edge bindings will also need updated.  So the interface for this endpoint should be the same as the /query interface from trapi.

**Comment by kshefchek** [comment]
> we can leave out the camel case stuff - it'll be better fixed upstream (babel).
This should be fixed with 49de5d0, although I updated the mock redis files to have camel case types so the unit test did not change

> A little bigger, but hopefully not too bad, I think what we would like here is a full TRAPI interface,
Will do! We did discuss this originally so this was an oversight on my part

**Comment by kshefchek** [comment]
I've update the code to accept and output TRAPI 1.0 messages.

Test files I'm using for the general case: https://github.com/kshefchek/NodeNormalization/tree/00c9ec/tests/resources

this testing approach will become unwieldy at some point - we should come back and write smaller tests for edge cases

**Comment by cbizon** [comment]
A couple of minor 1.0 issues,  I think.

First, nodes only have 'category', not 'type'.

Second, you can't have the "same_as" property directly on the node.  The Node can only have id, name, category and attributes.  Then attributes is a list of Attributes that follow this schema:
```
Attribute:
      type: object
      description: Generic attribute for a node
      properties:
        name:
          type: string
          description: >-
            Human-readable name or label for the attribute. Should be the name
            of the semantic type term.
          example: PubMed Identifier
        value:
          example: 32529952
          description: >-
            Value of the attribute. May be any data type, including a list.
        type:
          type: string
          description: >-
            CURIE of the semantic type of the attribute, from the EDAM ontology
            if possible. If a suitable identifier does not exist, enter a
            descriptive phrase here and submit the new type for consideration
            by the appropriate authority.
          example: EDAM:data_1187
        url:
          type: string
          description: >-
            Human-consumable URL to link out and read about the attribute (not
            the node).
          example: https://pubmed.ncbi.nlm.nih.gov/32529952
        source:
          type: string
          description: Source of the attribute, as a CURIE prefix.
          example: UniProtKB
      required:
        - type
        - value
```

So instead of 
```
{
      "id": "MONDO:0017148",
      "type": ["biolink:Disease"],
      "name": "heritable pulmonary arterial hypertension",
      "same_as": ["MONDO:0017148", "ORPHANET:275777", "UMLS:CN202575", "NCIT:C121945", "SNOMEDCT:697897003"],
      "category": ["biolink:Disease", "biolink:DiseaseOrPhenotypicFeature", "biolink:BiologicalEntity", "biolink:NamedThing"]
 }
```

it will be
```
{
      "id": "MONDO:0017148",
      "name": "heritable pulmonary arterial hypertension",
      "category": ["biolink:Disease", "biolink:DiseaseOrPhenotypicFeature", "biolink:BiologicalEntity", "biolink:NamedThing"]
      "attributes": [
          {
           "name": "same_as",
           "value": ["MONDO:0017148", "ORPHANET:275777", "UMLS:CN202575", "NCIT:C121945", "SNOMEDCT:697897003"],
           "type": "<i'm unsure what would go here>",
           "source": "NodeNormalizer"
        }
      ]
}
```

@patrickkwang is that right?

**Comment by cbizon** [comment]
Maybe it would be

```
{
  "name": "same as"
  "type": "biolink:same_as"
  ...
}

**Comment by kshefchek** [comment]
These are added as Attributes on the kgraph so they follow the attribute schema here - https://github.com/TranslatorSRI/reasoner-pydantic/blob/v1.0/reasoner_pydantic/shared.py#L10

kgraph node - https://github.com/TranslatorSRI/reasoner-pydantic/blob/v1.0/reasoner_pydantic/kgraph.py#L18

https://github.com/kshefchek/NodeNormalization/blob/00c9ec1b18aa2a63627dfbbf4027d97eb9158d9d/tests/resources/postmerged_message.json#L64

**Comment by cbizon** [comment]
Huh.  The output here: https://github.com/kshefchek/NodeNormalization/blob/00c9ec/tests/resources/postmerged_message.json#L64 looks right, so maybe you just pasted something old into the comment box above (which is what my comments were referring to)

**Comment by kshefchek** [comment]
oh yeah sorry - the json example in https://github.com/TranslatorSRI/NodeNormalization/pull/35#issuecomment-726395574 was reworked after moving to trapi 1.0

**Comment by cbizon** [comment]
Oh right, that's my bad. Sorry about that.  Looks good to me then.

**Comment by kshefchek** [comment]
the only part I don't feel 100% comfortable with is the edge merging, but was thinking we could come back to it (ie the criteria for a merge and edge cases)

File: Dockerfile [hunk 1]
```diff
@@ -18,11 +18,11 @@ RUN pip install -r /home/murphy/requirements.txt --src /usr/local/src
 # install library
 ADD ./r3 /home/murphy/r3
 ADD ./swagger_ui /home/murphy/swagger_ui
-ADD ./main.py /home/murphy/main.py
 ADD ./setup.py /home/murphy/setup.py
-ADD ./src /home/murphy/src
+ADD ./src /home/murphy/node_normalizer
```
Comment (YaphetKG) [inline Dockerfile:22]: Hi @kshefchek , I think this needs to be `ADD ./node_normalizer /home/murphy/node_normalizer`
also we need `ADD ./config.json /home/murphy/config.json`

File: Dockerfile [hunk 2]
```diff
@@ -18,11 +18,11 @@ RUN pip install -r /home/murphy/requirements.txt --src /usr/local/src
...
 
 RUN pip install -e .
 
 # setup entrypoint
-ENTRYPOINT ["python", "/home/murphy/main.py"]
\ No newline at end of file
+# gunicorn, hypercorn also options https://fastapi.tiangolo.com/deployment/manually/
+ENTRYPOINT ["python", "uvicorn node_normalizer.server:app --app-dir /home/murphy/ --port 6380"]
```
Comment (YaphetKG) [inline Dockerfile:29]: Python complains with no such file for this command, needs to be something like
`ENTRYPOINT ["python", "-m" , "uvicorn", "node_normalizer.server:app", "--app-dir", "/home/murphy/", "--port", "6380"]`

File: Dockerfile (hunk 1 above)
Comment (kshefchek) [reply Dockerfile]: done!

**Review by YaphetKG — CHANGES_REQUESTED** [review]
Hi Kent apologies (thought this was viewable to you on git), this was also giving me some issues,

File: Dockerfile (hunk 2 above)
Comment (kshefchek) [reply Dockerfile]: fixed!
